alf/algorithms/algorithm.py

```diff
@@ -150,43 +96,49 @@ def __init__(self,
             debug_summaries (bool): True if debug summaries should be created.
             name (str): name of this algorithm.
         """
-        super(Algorithm, self).__init__(name=name)
+        super(Algorithm, self).__init__()
 
+        self._name = name
         self._train_state_spec = train_state_spec
         if predict_state_spec is None:
             predict_state_spec = train_state_spec
         self._predict_state_spec = predict_state_spec
-        self._is_rnn = len(tf.nest.flatten(train_state_spec)) > 0
-
-        if not optimizer:
-            assert trainable_module_sets is None
-            self._init_optimizers = [None]
-            self._init_module_sets = [[]]
-        elif isinstance(optimizer, tf.optimizers.Optimizer):
-            assert trainable_module_sets is None
-            self._init_optimizers = [optimizer]
-            self._init_module_sets = [[]]
-        else:
-            assert isinstance(optimizer, list)
-            assert isinstance(trainable_module_sets, list)
-            assert len(trainable_module_sets) == len(optimizer), (
-                "`optimizer` and `trainable_module_sets`"
-                "should have same length")
-            self._init_optimizers = optimizer
-            self._init_module_sets = trainable_module_sets
-
-        self._cached_opt_and_var_sets = None
+        self._is_rnn = len(alf.nest.flatten(train_state_spec)) > 0
+
         self._gradient_clipping = gradient_clipping
         self._clip_by_global_norm = clip_by_global_norm
         self._debug_summaries = debug_summaries
+        self._default_optimizer = optimizer
+        self._optimizers = []
+        self._default_optimizer = optimizer
+        self._module_to_optimizer = {}
+        if optimizer:
+            self._optimizers.append(optimizer)
 
     @property
-    def trainable_variables(self):
-        """Returns all trainable variables including child algorithms/modules.
-        Some variables might be optimized by optimizers of child algorithms.
+    def name(self):
+        """The name of this algorithm."""
+        return self._name
+
+    def add_optimizer(self, optimizer: torch.optim.Optimizer,
+                      modules_and_params):
+        """Add an optimizer
+
+        Note that the modules and params contained in `modules_and_params`
+        should still be the attributes of the algorithm (i.e., they can be
+        retrieved in self.children() or self.parameters())
+
+        Args:
+            optimizer (Optimizer): optimizer
+            modules_and_params (list of Module or Parameter): The modules and
+                parameters to be optimized by `optimizer`
         """
-        return sum([var_set for _, var_set in self._get_opt_and_var_sets()],
-                   [])
+        params = []
```

Comment:
`params` not used anywhere in this function?

Reply:
fixed